### PR TITLE
Donation Spending Breakdowns

### PR DESCRIPTION
**Pull request by sampoder**
Context: https://hackclub.slack.com/archives/CN523HLKW/p1697745184336139

Example: [breakdown.pdf](https://github.com/hackclub/hcb/files/13895171/breakdown.pdf)

For now, I've only made them accessible to admins but it'd be interesting to expand that.

### CODE REVIEW

File: app/controllers/hcb_codes_controller.rb [hunk 1]
```diff
@@ -168,4 +168,120 @@ def toggle_tag
     redirect_back fallback_location: @event
   end
 
+  def breakdown
+    @hcb_code = HcbCode.find_by(hcb_code: params[:id]) || HcbCode.find(params[:id])
+    authorize @hcb_code
+
+    if @hcb_code.amount_cents <= 0
+      return redirect_to @hcb_code
+    end
```
Comment (garyhtou) [inline app/controllers/hcb_codes_controller.rb]:
```suggestion
    redirect_to @hcb_code and return if @hcb_code.amount_cents <= 0
```

File: config/routes.rb [hunk 2]
```diff
@@ -333,6 +333,7 @@
       get "attach_receipt"
       get "memo_frame"
       get "dispute"
+      get "breakdown", to: "hcb_codes#breakdown", as: :breakdown
```
Comment (garyhtou) [inline config/routes.rb]: i think we may be able to move this route in here
https://github.com/hackclub/bank/blob/604c5197cc8764f7b07af5aecb4d6435a528651e/config/routes.rb#L331-L331

File: app/views/hcb_codes/show.html.erb [hunk 3]
```diff
@@ -2,6 +2,12 @@
 
 <%= render "events/nav" if @event %>
 
+<% admin_tools("mt3") do %>
+  <% if @hcb_code.amount_cents > 0 %>
```
Comment (garyhtou) [inline app/views/hcb_codes/show.html.erb]:
```suggestion
  <% if @hcb_code.amount_cents.positive? %>
```

File: app/controllers/hcb_codes_controller.rb [hunk 4]
```diff
@@ -168,4 +168,120 @@ def toggle_tag
...
+        event
+      rescue
+        @hcb_code.events.min_by do |e|
+          [e.users.include?(current_user), e.is_public?].map { |b| b ? 0 : 1 }
+        end
+      rescue
+        @hcb_code.event
+      end
+
+    @donations = {}
```
Comment (garyhtou) [inline app/controllers/hcb_codes_controller.rb]: Can we move this logic into a service? this way we can break each section into it's own method to make it more readable

**Review by garyhtou — CHANGES_REQUESTED** [review]

**Comment by sampoder** [comment]
Alrightly, I cleaned up a couple of things and moved stuff into a service! I choose to use an EventService as it felt like that made more sense. @garyhtou thoughts?

File: app/controllers/hcb_codes_controller.rb [hunk 5]
```diff
@@ -168,4 +168,64 @@ def toggle_tag
     redirect_back fallback_location: @event
   end
 
+  def breakdown
+    @hcb_code = HcbCode.find_by(hcb_code: params[:id]) || HcbCode.find(params[:id])
+    authorize @hcb_code
+
+    if @hcb_code.amount_cents <= 0
+      return redirect_to @hcb_code
+    end
```
Comment (garyhtou) [inline app/controllers/hcb_codes_controller.rb]:
```suggestion
    return redirect_to @hcb_code unless @hcb_code.amount_cents.positive?
```

File: app/controllers/hcb_codes_controller.rb [hunk 6]
```diff
@@ -168,4 +168,64 @@ def toggle_tag
...
+        raise StandardError unless @hcb_code.events.include? event
+
+        event
+      rescue
+        @hcb_code.events.min_by do |e|
+          [e.users.include?(current_user), e.is_public?].map { |b| b ? 0 : 1 }
+        end
+      rescue
+        @hcb_code.event
+      end
```
Comment (garyhtou) [inline app/controllers/hcb_codes_controller.rb]:
```suggestion
    @event = @hcb_code.event
```

File: app/views/hcb_codes/breakdown.pdf.erb
Comment (garyhtou) [inline app/views/hcb_codes/breakdown.pdf.erb:1]: Add Sarthak's emoji stuff

File: app/services/event_service/pair_income_with_spending.rb [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+      end
+      @income
+    end
+
+    def distribute_spending_to_income(ct:)
+      distributed = 0
+      while distributed < ct.amount_cents.abs
+        available_on_current = current[:available]
+        # the piece of income at the current index has enough space for the rest that needs to be distributed.
+        if available_on_current > (ct.amount_cents.abs - distributed)
```
Comment (garyhtou) [inline app/services/event_service/pair_income_with_spending.rb]: Can we give `ct.amount_cents.abs - distributed` it's own variable?

File: app/services/event_service/pair_income_with_spending.rb [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+            id: ct.id,
+            memo: ct.local_hcb_code.memo,
+            amount: ct.amount_cents,
+            spent_on: [],
+            available: ct.amount_cents
+          }
+        else
+          distribute_spending_to_income(ct:)
+        end
+      end
```
Comment (garyhtou) [inline app/services/event_service/pair_income_with_spending.rb]: this should handle the case where the event was negative in the past

```suggestion
      negative = @event.canonical_transactions.order("date asc").filter_map do |ct|
        if ct.amount_cents > 0
          @income[ct.id.to_s] = {
            id: ct.id,
            memo: ct.local_hcb_code.memo,
            amount: ct.amount_cents,
            spent_on: [],
            available: ct.amount_cents
          }
          next
        end

        ct
      end

      negative.each do |ct|
        next unless @income.keys[@current_index]
        distribute_spending_to_income(ct:)
      end
```

File: app/controllers/hcb_codes_controller.rb [hunk 9]
```diff
@@ -231,4 +231,64 @@ def invoice_as_personal_transaction
...
+
+    @income = ::EventService::PairIncomeWithSpending.new(event: @event).run
+
+    @spent_on = []
+    @available = 0
+
+    @hcb_code.canonical_transactions.each do |ct|
```
Comment (garyhtou) [inline app/controllers/hcb_codes_controller.rb]:
```suggestion
    # PairIncomeWithSpending is done on a per CanonicalTransaction basis
    # This compute it for this specific HcbCode
    @hcb_code.canonical_transactions.each do |ct|
```

File: app/controllers/hcb_codes_controller.rb [hunk 10]
```diff
@@ -231,4 +231,64 @@ def invoice_as_personal_transaction
...
+      if (ct[:amount_cents] > 0) && @income[ct[:id].to_s]
+        @spent_on.concat @income[ct[:id].to_s][:spent_on]
+        @available += @income[ct[:id].to_s][:available]
+      end
+    end
+
+    @spent_on = @spent_on.group_by { |hash| hash[:memo] }.map do |memo, group|
+      total_amount = group.reduce(0) { |sum, item| sum + item[:amount] }
```
Comment (garyhtou) [inline app/controllers/hcb_codes_controller.rb]:
```suggestion
      total_amount = group.sum(0) { |item| item[:amount] }
```

File: app/views/hcb_codes/breakdown.pdf.erb [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    
+  </style>
+</head>
+<body>
+  <br />
+  <div class="container">
+    <br />
+    <p style="font-size: 12.5pt;"><span class="bold"><%= @event.name %> Donation Breakdown</span></p>
+    <br />
+    <% if @spent_on.length > 0 %>
```
Comment (garyhtou) [inline app/views/hcb_codes/breakdown.pdf.erb]:
```suggestion
    <% if @spent_on.any? %>
```

File: app/views/hcb_codes/breakdown.pdf.erb [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+        <% if  @event.is_public?  %>
+          <p>
+            Additionally, <%= @event.name %> have made their entire finances transparent at <a href="https://hcb.hackclub.com/<%= @event.slug %>">hcb.hackclub.com/<%= @event.slug %></a>.
+          </p>
+        <% end %>
+      </div>
+    <% else %>
+      <div style="line-height: 1.5;">
+        <p>
+          Thank you for your kind donation towards <%= @event.name %>, your support is truly appreciated. At the moment, all of your donation is yet to be spent.
```
Comment (garyhtou) [inline app/views/hcb_codes/breakdown.pdf.erb]:
```suggestion
          Thank you for your kind donation towards <%= @event.name %>, your support is truly appreciated. At the moment, your donation has yet to be spent.
```

File: app/views/hcb_codes/breakdown.pdf.erb [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+      <div style="line-height: 1.5;">
+        <p>
+          Thank you for your kind donation towards <%= @event.name %>, your support is truly appreciated. This document serves as a breakdown of how your donation was spent.
+        </p>
+        <table style="line-height: 1.15">
+          <tbody>
+            <% @spent_on.each do |transaction| %>
+              <tr>
+                <td class="bold" colspan="2">
+                  <%= transaction[:memo].gsub(/[^[:alnum:][:blank:][:punct:]]/, '').squeeze(' ').strip %>
```
Comment (garyhtou) [inline app/views/hcb_codes/breakdown.pdf.erb]:
```suggestion
                  <%= link_to transaction[:memo].strip, transaction[:url] %>
```

File: app/services/event_service/pair_income_with_spending.rb [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+    def distribute_spending_to_income(ct:)
+      distributed = 0
+      while distributed < ct.amount_cents.abs
+        available_on_current = current[:available]
+        # the piece of income at the current index has enough space for the rest that needs to be distributed.
+        if available_on_current > (ct.amount_cents.abs - distributed)
+          current[:spent_on].append(
+            {
+              id: ct.id,
+              memo: ct.local_hcb_code.memo,
```
Comment (garyhtou) [inline app/services/event_service/pair_income_with_spending.rb]:
```suggestion
              memo: ct.local_hcb_code.memo,
              url: ct.local_hcb_code.url,
```

File: app/controllers/hcb_codes_controller.rb [hunk 15]
```diff
@@ -231,4 +231,64 @@ def invoice_as_personal_transaction
...
+      { id: group.first[:id], memo:, amount: total_amount }
```
Comment (garyhtou) [inline app/controllers/hcb_codes_controller.rb]:
```suggestion
      significant_transaction = group.max_by { |t| t[:amount] }
      { id: significant_transaction[:id], memo:, amount: total_amount, url: significant_transaction[:url] }
```

**Review by garyhtou — APPROVED** [review]
Great!

In the future:
- Send via email for large events (e.g. HQ)
- Improve performance of `PairIncomeWithSpending` by using a running total

**Pull request merged**